### PR TITLE
[Core] Support setting node labels for ray up

### PR DESCRIPTION
**Pull request by jjyao**
## Why are these changes needed?
Support setting node labels in cluster.yaml

The code mirrors the setting node custom resources code path.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/autoscaler/_private/updater.py [hunk 1]
```diff
@@ -509,6 +515,15 @@ def do_update(self):
                     else:
                         pass
 
+                    # Add a label override env variable if needed:
```
Comment (larrylian) [inline python/ray/autoscaler/_private/updater.py:518]:
```
if self.provider_type != "local":
      if self.node_resources:
            env_vars[RESOURCES_ENVIRONMENT_VARIABLE] = self.node_resources
      if self.node_labels:
             env_vars[LABELS_ENVIRONMENT_VARIABLE] = self.node_labels
```

File: python/ray/autoscaler/v2/instance_manager/ray_installer.py [hunk 2]
```diff
@@ -80,6 +80,7 @@ def install_ray(self, instance: Instance, head_node_ip: str) -> bool:
             use_internal_ip=True,
             docker_config=docker_config,
             node_resources=self._config.get_node_resources(instance.instance_type),
+            node_labels=self._config.get_node_labels(instance.instance_type),
```
Comment (rickyyx) [inline python/ray/autoscaler/v2/instance_manager/ray_installer.py:83]: So we are assuming labels are defined on per isntance type granularity?

I wonder if per-instance (maybe in the future) would make more sense?

Comment (rickyyx) [reply python/ray/autoscaler/v2/instance_manager/ray_installer.py:83]: but i guess we don't have any per instance concept in autoscaler spec yet. Or maybe per instance type is good already?

File: python/ray/tests/test_advanced_8.py [hunk 3]
```diff
@@ -192,6 +192,18 @@ def test_ray_resources_environment_variable(ray_start_cluster):
     assert cluster_resources["CPU"] == 3
 
 
+def test_ray_labels_environment_variables(shutdown_only):
```
Comment (rickyyx) [inline python/ray/tests/test_advanced_8.py:195]: Could we also add a small test for https://github.com/ray-project/ray/blob/0e0be3a86e3a608bf9418bbedce5d2e74cf2953a/python/ray/util/state/common.py#L511-L512 ?

**Review by rickyyx — APPROVED** [review]
LGTM - mostly nits!

File: python/ray/_private/ray_constants.py [hunk 4]
```diff
@@ -174,6 +174,7 @@ def env_set_by_user(key):
 RESOURCE_CONSTRAINT_PREFIX = "accelerator_type:"
 
 RESOURCES_ENVIRONMENT_VARIABLE = "RAY_OVERRIDE_RESOURCES"
+LABELS_ENVIRONMENT_VARIABLE = "RAY_OVERRIDE_LABELS"
```
[the remark this answers is not in the document]
Comment (jjyao) [reply python/ray/_private/ray_constants.py:177]: I'm trying to mirror `RESOURCES_ENVIRONMENT_VARIABLE`: the override means that if you set labels both via `ray start` and cluster.yaml, the latter will override the former. There is actually a test for it.

File: python/ray/tests/test_advanced_8.py [hunk 5]
```diff
@@ -178,11 +178,11 @@ def test_ray_address_environment_variable(ray_start_cluster):
     ray.shutdown()
 
 
-def test_ray_resources_environment_variable(ray_start_cluster):
-    address = ray_start_cluster.address
-
-    os.environ["RAY_OVERRIDE_RESOURCES"] = '{"custom1":1, "custom2":2, "CPU":3}'
-    ray.init(address=address, resources={"custom1": 3, "custom3": 3})
+def test_ray_resources_environment_variable(shutdown_only):
```
[the remark this answers is not in the document]
Comment (jjyao) [reply python/ray/tests/test_advanced_8.py:181]: Not really a bug but unnecessary code.

File: python/ray/autoscaler/v2/instance_manager/ray_installer.py (hunk 2 above)
Comment (jjyao) [reply python/ray/autoscaler/v2/instance_manager/ray_installer.py:83]: Yea. It's the same as resources, there is only per instance type granularity not per instance. Although you can create an node type for each instance.

File: python/ray/_private/ray_constants.py (hunk 4 above)
Comment (rkooo567) [inline python/ray/_private/ray_constants.py:177]: We should comment how to use this env var?

File: python/ray/_private/node.py [hunk 6]
```diff
@@ -415,6 +415,38 @@ def _init_temp(self):
...
+
+        env_labels = {}
+        env_string = os.getenv(ray_constants.LABELS_ENVIRONMENT_VARIABLE)
+        if env_string:
+            try:
+                env_labels = json.loads(env_string)
+            except Exception:
+                logger.exception(f"Failed to load {env_string}")
+                raise
+            logger.debug(f"Autoscaler overriding labels: {env_labels}.")
```
Comment (rkooo567) [inline python/ray/_private/node.py:444]:
```suggestion
            logger.info(f"Autoscaler overriding labels: {env_labels}.")
```
Seems it is worth info?

File: python/ray/_private/node.py [hunk 7]
```diff
@@ -415,6 +415,38 @@ def _init_temp(self):
...
+
+        return merge_labels(
+            env_labels, self._ray_params.labels if self._ray_params.labels else {}
```
Comment (rkooo567) [inline python/ray/_private/node.py:447]:
```suggestion
            env_labels, self._ray_params.labels or {}
```
I believe it is possible?

File: python/ray/autoscaler/_private/autoscaler.py [hunk 8]
```diff
@@ -1018,6 +1020,13 @@ def _node_resources(self, node_id):
         else:
             return {}
 
+    def _node_labels(self, node_id):
```
Comment (rkooo567) [inline python/ray/autoscaler/_private/autoscaler.py:1023]: Maybe node_labels_from_config? Can this return labels passed by the env var?

File: python/ray/autoscaler/_private/updater.py [hunk 9]
```diff
@@ -509,6 +515,15 @@ def do_update(self):
...
+                    if self.provider_type == "local":
+                        # Local NodeProvider doesn't need label override.
+                        pass
+                    elif self.node_labels:
+                        env_vars[LABELS_ENVIRONMENT_VARIABLE] = self.node_labels
```
Comment (rkooo567) [inline python/ray/autoscaler/_private/updater.py:523]: Q: no reason to validate schema (or is it already done before coming here?)

File: python/ray/tests/test_advanced_8.py [hunk 10]
```diff
@@ -192,6 +192,18 @@ def test_ray_resources_environment_variable(ray_start_cluster):
     assert cluster_resources["CPU"] == 3
 
 
+def test_ray_labels_environment_variables(shutdown_only):
+    os.environ[
+        ray_constants.LABELS_ENVIRONMENT_VARIABLE
+    ] = '{"custom1":"1", "custom2":"2"}'
+    ray.init(labels={"custom1": "3", "custom3": "3"})
```
Comment (rkooo567) [inline python/ray/tests/test_advanced_8.py:199]: Hmm should we support labels from `ray.init`? I am not sure if this was already discussed, but it seems not useful. If it is a single node, ability to pass label is not necessary?

**Review by rkooo567 — APPROVED** [review]

File: python/ray/tests/test_advanced_8.py (hunk 10 above)
Comment (jjyao) [reply python/ray/tests/test_advanced_8.py:199]: Yea, you are right it's not useful in practice but it does make things symmetric (you can set node labels in all ways of starting a node). After all, users can always use `ray start` to start a single node cluster with node labels.

It's also convenient for users to try labels out.

File: python/ray/_private/node.py [hunk 11]
```diff
@@ -415,6 +415,38 @@ def _init_temp(self):
         )
         try_to_create_directory(self._runtime_env_dir)
 
+    def _get_node_labels(self):
+        def merge_labels(env_dict, params_dict):
```
[the remark this answers is not in the document]
Comment (jjyao) [reply python/ray/_private/node.py:419]: rename to `env_override_dict`

File: python/ray/autoscaler/_private/autoscaler.py (hunk 8 above)
Comment (jjyao) [reply python/ray/autoscaler/_private/autoscaler.py:1023]: No. This is from cluster.yaml. I name it this way to be consistent with `_node_resources`

File: python/ray/_private/node.py [hunk 12]
```diff
@@ -415,6 +415,38 @@ def _init_temp(self):
...
+                        "Autoscaler is overriding your label:"
+                        f"{key}: {params_dict[key]} with {env_dict[key]}."
+                    )
+            return result
+
+        env_labels = {}
+        env_string = os.getenv(ray_constants.LABELS_ENVIRONMENT_VARIABLE)
+        if env_string:
+            try:
+                env_labels = json.loads(env_string)
```
[the remark this answers is not in the document]
Comment (jjyao) [reply python/ray/_private/node.py:442]: We have already done the validation when parsing cluster.yaml so we don't need to do it again here. Also parse_node_labels_json requires cli_logger and the error message is tailored for `ray start`

**Pull request merged**